aubio-rs/src/vec.rs

```diff
@@ -324,3 +326,163 @@ impl<'a, T: AsMut<[f32]>> From<T> for CVecPhasMut<'a> {
         }
     }
 }
+
+/**
+ * Immutable matrix of real valued data.
+ */
+#[repr(transparent)]
+pub struct FMat<'a> {
+    fmat: ffi::fmat_t,
+    _pd: PhantomData<&'a ()>,
+}
+
+impl<'a> FMat<'a> {
+    pub(crate) fn as_ptr(&'a self) -> *const ffi::fmat_t {
+        &self.fmat
+    }
+
+    pub fn length(&self) -> usize {
+        self.fmat.length as usize
+    }
+
+    pub fn height(&self) -> usize {
+        self.fmat.height as usize
+    }
+
+    pub fn print(&self) {
+        unsafe { ffi::fmat_print(self.as_ptr()) }
+    }
+
+    /**
+     * Create a matrix from an already existing `fmat_t` pointer.
+     *
+     * The matrix is non-owned; useful to avoid double-frees for `FilterBank`,
+     * for instance.
+     */
+    pub unsafe fn from_raw_ptr<'b>(ptr: *const ffi::fmat_t) -> FMat<'b> {
+        FMat {
+            fmat: *ptr,
+            _pd: PhantomData,
+        }
+    }
+
+    /// Read sample value in a buffer 
+    pub fn get_sample(&self, channel: usize, position: usize) -> Result<f32> {
+        if channel >= self.height() || position >= self.length() {
+            return Err(Error::InvalidArg);
+        }
+        Ok(unsafe {
+            ffi::fmat_get_sample(
+                self.as_ptr(),
+                channel as ffi::uint_t,
+                position as ffi::uint_t,
+            )
+        })
+    }
+
+    pub fn get_vec(&self) -> Vec<&mut [f32]> {
+        let mut vec = Vec::with_capacity(self.height());
+        let mut ptr = self.fmat.data;
+        let end = self.fmat.data.wrapping_add(self.height());
+
+        while ptr != end {
+            vec.push(unsafe { slice::from_raw_parts_mut(*ptr, self.length()) });
+            ptr = ptr.wrapping_add(1);
+        }
+        vec
+    }
+
+    fn from_wrapper<T: AsRef<Vec<&'a [f32]>>>(data: T) -> Result<Self> {
+        let data = data.as_ref();
+        let height = data.len() as ffi::uint_t;
+        let length = data[0].len();
+        for x in data.iter() {
+            if x.len() != length {
+                return Err(Error::FailedInit);
+            }
+        }
+        // TODO leaking memory here!
+        let data = Box::into_raw(
+            data.iter()
+                .map(|s| s.as_ptr() as *mut _)
+                .collect::<Vec<*mut f32>>()
+                .into_boxed_slice(),
+        );
```

Comment:
I suppose you does this thing wrong.
Actually, we don't need copy vector, we can just transmute it.
I mean the `Vec<&T>` is same as `Vec<T*>` in terms of data mapping.